Catch duplicate SSN attempt during IdV

pkarman opened the pull request:
**Why**: SSN must be unique, so alert the user
as early as possible of the problem.

sbc100 commented on app/controllers/idv/sessions_controller.rb:
```diff
@@ -31,6 +36,19 @@ def update
     private
 
+    def validation_errors
+      @_validation_errors ||= []
```
Is this pattern of using `_` a common thing in ruby / rails?  Should we be using it more?  Does it have actual meaning in ruby?

pkarman replied:
I have seen it done for "private" variables as a way of saying "hands off for direct access". It's only a convention and doesn't mean anything to the compiler, only to the programmer.

monfresh replied:
I would recommend using a Form Object and Validator for this logic as opposed to putting it in the controller. See our `/forms` and `/validators` directories for examples.

monfresh replied:
As this article says, "keep business logic out of the controller and views": https://robots.thoughtbot.com/activemodel-form-objects

sbc100 commented on config/locales/en.yml:
```diff
@@ -147,6 +147,8 @@ en:
       state: State
       zipcode: ZIP Code
+    errors:
+      duplicate_ssn: 'You may already have an account, please use that one. You can always change your email address in your existing account.'
```
Perhaps more explicit?  `An account has already been created using this SSN.`?  I guess the wording can be tweaked later.

sbc100 replied:
We tend not to use quotes in this file.  Do you need to here?   Also can you split this onto multiple lines?

sbc100 commented:
lgtm % comments

pkarman commented:
locale string came directly from the UX error document (in process). I have removed the quotation marks and split into two lines. If that is really the convention there are several instances that should be cleaned up.

sbc100 commented:
Yes, en.yml needs a cleanup.

sbc100 commented:
lgtm

pkarman commented:
thanks @monfresh -- I will revisit.

pkarman commented:
ok @monfresh -- refactored to use a Form object

sbc100 commented on app/forms/idv_profile_form.rb:
```diff
@@ -0,0 +1,48 @@
+    @params = params
+    error_required_params if required_params_blank?
+    error_duplicate_ssn if ssn_taken?
+    self
```
Is returning self a normal thing to do in a submit method?  Other places I see `submit` return true and success and false on failure.

pkarman replied:
I will confess I did that as an afterthought to make it easier to get the one-liner in the controller. I will change.

pkarman commented:
all 💚  again

sbc100 commented on app/views/shared/_messages.html.slim:
```diff
@@ -1,10 +1,12 @@
-    - if msg.present? && msg.is_a?(String)
+  - flash.each do |name, message|
+    - if message.present? && (message.is_a?(String) || message.is_a?(Array))
+      - messages = message.is_a?(Array) ? message : [message]
```
@monfresh Is this a sensible approach?

sbc100 commented:
lgtm, but I'll lat @monfresh have the final say (specifically regarding the error list thing which I'm not sure about)

monfresh commented:
@pkarman I just pushed a second commit that cleans things up. Let me know what you think.

monfresh commented:
@sbc100 can you take a look at my commit, please?

sbc100 commented:
lgtm... do you want to submit both CLs are part of this one PR, or have peter squash your change into his?

monfresh commented:
It might be worth preserving both commits so we can refer to this in the future, for example in a blog post that shows before and after.
